jetty-util/src/main/java/org/eclipse/jetty/util/thread/ReservedThreadExecutor.java

```diff
@@ -266,139 +265,144 @@ public String toString()
+                    // Always poll at some period as safety to ensure we don't poll forever.
+                    Runnable task = _idleTime <= 0 ? _queue.poll(30, TimeUnit.SECONDS) : _queue.poll(_idleTime, _idleTimeUnit);
                     if (LOG.isDebugEnabled())
-                        LOG.debug("{} task={}", this, task);
+                        LOG.debug("{} task={} {}", this, task, ReservedThreadExecutor.this);
                     if (task != null)
                         return task;
 
-                    if (_stack.remove(this))
+                    // Has the RTE stopped?
+                    if (_count.getLo() < 0)
+                        return STOP;
+
+                    // Have we idled out?
+                    if (_idleTime > 0)
                     {
-                        if (LOG.isDebugEnabled())
-                            LOG.debug("{} IDLE", this);
-                        _size.decrementAndGet();
+                        _state = State.IDLE;
+                        _count.add(0, -1);
                         return STOP;
                     }
                 }
                 catch (InterruptedException e)
                 {
                     LOG.ignore(e);
-                    // If the wait was interrupted, then STOP if we are not running
-                    if (!isRunning())
-                        return STOP;
                 }
             }
         }
 
         @Override
         public void run()
         {
-            while (isRunning())
+            _thread = Thread.currentThread();
+            try
             {
-                // test and increment size BEFORE decrementing pending,
-                // so that we don't have a race starting new pending.
-                int size = _size.get();
+                while (true)
+                {
+                    long count = _count.get();
 
-                // Are we stopped?
-                if (size < 0)
-                    return;
+                    // reduce pending if this thread was pending
+                    int pending = getHi(count) - (_state == State.PENDING ? 1 : 0);
+                    int size = getLo(count);
 
-                // Are we surplus to capacity?
-                if (size >= _capacity)
-                {
-                    if (LOG.isDebugEnabled())
-                        LOG.debug("{} size {} > capacity", this, size, _capacity);
-                    if (_starting)
-                        _pending.decrementAndGet();
-                    return;
-                }
+                    State next;
+                    if (size < 0 || size >= _capacity)
+                    {
+                        // The executor has stopped or this thread is excess to capacity
+                        next = State.STOPPED;
+                    }
+                    else if (size > 0 && _idleTime > 0 && _idleTimeUnit.toNanos(_idleTime) < (System.nanoTime() - _lastZero))
```

Comment:
Okay, you're kinda sorta reproducing the previous "stack" effect by expiring reserved threads. Got it.

Reply:
I would prefer this update to be atomic using `AtomicLong.updateAndGet()`, otherwise many threads can see the expression to be true and idle themselves.